src/raftstore/coprocessor/split_check/table.rs

```diff
@@ -0,0 +1,519 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+
+use std::cmp::Ordering;
+use rocksdb::{SeekKey, DB};
+use kvproto::metapb::Region;
+
+use storage::CF_WRITE;
+use raftstore::store::keys;
+use raftstore::store::engine::{IterOption, Iterable};
+use coprocessor::codec::table as table_codec;
+
+use super::super::{Coprocessor, ObserverContext, RegionObserver, Result};
+use super::Status;
+
+#[derive(Default)]
+pub struct TableStatus {
+    first_encoded_table_prefix: Vec<u8>,
+    last_encoded_table_prefix: Option<Vec<u8>>,
+}
+
+#[derive(Default)]
+pub struct TableCheckObserver;
+
+impl Coprocessor for TableCheckObserver {}
+
+impl RegionObserver for TableCheckObserver {
+    fn new_split_check_status(&self, ctx: &mut ObserverContext, status: &mut Status, engine: &DB) {
+        let mut table_status = TableStatus::default();
+        let skip = before_check(&mut table_status, engine, ctx.region());
+        if !skip {
+            status.table = Some(table_status);
+        }
+    }
+
+    fn on_split_check(
+        &self,
+        _: &mut ObserverContext,
+        status: &mut Status,
+        key: &[u8],
+        _: u64,
+    ) -> Option<Vec<u8>> {
+        if let Some(ref mut table_status) = status.table {
+            check_key(table_status, key)
+        } else {
+            None
+        }
+    }
+}
+
+/// Do some quick checks, true for skipping `check_key`.
+fn before_check(status: &mut TableStatus, engine: &DB, region: &Region) -> bool {
+    let (actual_start_key, actual_end_key) = match bound_keys(engine, region) {
```

Comment:
you don't need to access RocksDB every time. You can use region Range to check in memory at first.

If the Region start and end key have the same table prefix, you can return directly.